feat: 플레이어 오브젝트 애니메이션 구현

j25nkh opened the pull request:
## 개요
- 기존 플레이어 오브젝트는 rigging (관절모션)이 없는 형태였습니다.
- 단조로운 플레이를 넘어 UX를 향상시키기 위해서는 여러 형태의 모션을 취할 수 있는 플레이어 오브젝트의 필요성이 시급했습니다.
- 스테이지 로드시 플레이어 오브젝트의 모션은 가만히 서있는 포즈의 standing animation입니다.
- 키보드 입력시 플레이어 오브젝트의 모션은 걷는 포즈의 walking animation입니다.
- 플레이어 오브젝트가 바라보는 방향에 따라 전진 방향이 달라지는 로직을 활용하여 키컨트롤을 구현하였습니다.

## 작업 내용


https://user-images.githubusercontent.com/105766632/226881601-0e52a0e2-047d-4e32-920f-c0a6fa1f71bf.mov



### 파일형식 선정
- `react-three/drei` 라이브러리를 활용하여 3d 오브젝트를 켄버스에 구현할 때 선택할 수 있는 파일형식은 `gltf`와 `glb` 형식이 있습니다.
- 본 프로젝트에서는 3d 오브젝트의 애니메이션 정보가 `array`의 형태로 저장되어 있는 `glb` 파일형식이 더 적합하다고 판단하였습니다.

### 3D 오브젝트의 선정
- 본 프로젝트의 모든 스테이지를 관통하는 개념인 '착시효과를 통한 패스의 연결'을 실현하기 위해 위해 각 스테이지는 정육면체의 연결을 통해 구성하고 있습니다.
- 따라서 맵과 위화감이 없으면서도 충분히 low-poly여서 렌더링이 저하되지 않는 minecraft의 steve 캐릭터를 선정하였습니다.

### Rigging 애니메이션 정보 저장
- 3D 오브젝트의 애니메이션을 생성하기 위해 `blender` 프로그램을 사용하여 `fbx` 형태의 오브젝트에 `rigging`을 구현하였습니다.
- `rigging`이 완료된 3D 오브젝트의 애니메이션을 key note를 찍는 방식으로 구현하였으며, 이를 각 `standingCycle` 및 `walkingCycle`에 저장하였습니다.
- 모션 애니메이션 정보가 생성된 오브젝트는 재질을 매핑하여 `glb`의 형태로 저장하였습니다.
![image](https://user-images.githubusercontent.com/105766632/226688446-52c3dea3-3851-4d52-ac58-f893ca2aac7f.png)

### 3D 오브젝트 불러오기
- `react-three/drei`의  `useGLTF`와 `useAnimations`을 사용하여 해당 오브젝트를 캔버스로 `import` 하였습니다.
- 불러온 오브젝트 애니메이션들은 각 모션별로 `fade in`, `fade out` 하는 방식으로 모션간 트랜지션을 주어 자연스럽게 연결하였습니다.

### 플레이어 오브젝트 머리 방향에 따른 전진축 변경로직
- ArrowUp keydown 이벤트의 경우 플레이어 오브젝트가 바라보는 방향에 따라 4가지 경우중 하나로 플레이어 오브젝트의 현재 위치를 업데이트 시켜줍니다 (x축, z축, -x축, -z축)
- 방향 전환 횟수를 추적하여, 아래의 배열에서 전진할 방향을 선택하는  로직을 사용하였습니다.
 ```
  const playerObjectHeadDirections = [
    { x: 0, z: 1 },
    { x: 1, z: 0 },
    { x: 0, z: -1 },
    { x: -1, z: 0 },
  ];
```
- 카운트를 4로 나눈 나머지값 번째의 배열을 다음 전진방향으로 설정하였으며, 화살표 좌클릭은 카운트 + 1, 우클릭은 카운트 +3으로, 우클릭시 배열을 역으로 순회하는 효과를 주었습니다.
```
  setPlayerObjectPosition(position => [
        position[0] -
          MOVEMENT_INCREMENT *
            playerObjectHeadDirections[rotationCount % 4].x,
        position[1],
        position[2] -
          MOVEMENT_INCREMENT *
            playerObjectHeadDirections[rotationCount % 4].z,
      ]);
```

## 특이 사항
- 키를 꾹 누르고 있는 상황에서 모션이 부자연스럽게 전환되는 현상이 있습니다. 
- 패스 자동감지 로직과 연결 후에도 해당 현상이 지속되는지 확인이 필요합니다.

j25nkh committed:
feat: 플레이어 오브젝트 애니메이션 구현

- 스테이지 로드시 플레이어 오브젝트 standing 형태로 구현
- 키보드 입력시 플레이어 오브젝트 walking animation 구현
- 플레이어 오브젝트의 헤드 방향에 따른 모션컨트롤 기능 구현

xollee commented on src/components/objects/PlayerObject.js:
```diff
@@ -1,23 +1,44 @@
 
-export default function PlayerObject({ position, rotation, scale }) {
-  const gltf = useLoader(GLTFLoader, "/assets/glb/playerObject.glb");
+export default function PlayerObject(props) {
```
props 수정해주심 좋을 것 같아요!

sageherb reviewed (dismissed):
고생하셨습니다 👏 오브젝트 걷는 모습이 귀엽네요
코드 조금만 다듬어주시면 좋을 것 같습니다!

j25nkh committed:
fix: Mock data 구조 수정

- 기존 array의 나열 방식에서 key와 coordinate 속성을 가진 객체의 나열로 변경
- 이에 따른 스테이지별 JSON파일 mapping 방식 수정

sageherb commented on src/components/objects/PlayerObject.js:
```diff
@@ -1,23 +1,133 @@
+  const { nodes, materials, animations } = useGLTF("/assets/glb/playerObj.glb");
+  const { actions, names } = useAnimations(animations, group);
+
+  const MOVEMENT_INCREMENT = 0.05;
+  const ROTATION_INCREMENT = (5 * Math.PI) / 180;
+  const KEYBOARD_EVENT_CODE = {
+    ARROW_UP: "ArrowUp",
+    ARROW_DOWN: "ArrowDown",
+    ARROW_LEFT: "ArrowLeft",
+    ARROW_RIGHT: "ArrowRight",
+  };
+  const playerObjectHeadDirections = [
+    { x: 0, z: 1 },
+    { x: 1, z: 0 },
+    { x: 0, z: -1 },
+    { x: -1, z: 0 },
+  ];
```
상수는 컴포넌트 바깥 위에 위치하면 좋을 것 같습니다
그리고 범용적으로 쓰이는 상수라면, 따로 파일을 분리해도 좋을 것 같아요

j25nkh replied:
해당 컴모넌트에서만 사용되는 상수들로 해당 파일 내 컴포넌트 바깥으로 위치 변경하였습니다.

yourchaee reviewed (dismissed):
가장 시간도 오래 소요되고 어려운 부분인데 부드럽게 움직이도록 구현하시느라 고생하셨습니다!
수고 많으셨습니다 :)

sageherb commented on src/components/objects/PlayerObject.js:
```diff
@@ -1,23 +1,133 @@
+          }, i * 10);
+        }
+      }
+    }, 180);
```
값만 바뀌고 반복되는 로직이 많아 보여요. 하나로 로직 정리가 가능할 것 같습니다.
그리고 PlayerObject 컴포넌트 안에서 움직이는 로직 부분이 있는 게 조금 어색합니다.
파일 분리를 해도 좋을 것 같아요.

j25nkh replied:
각 케이스별로 다수의 상태가 없데이트되어 일관된 로직을 찾아 정리하기 어려워 보입니다.
리뷰 내용을 반영하여 키컨트롤 로직을 별도의 파일로 분리하였습니다 :)

sageherb commented on src/components/objects/PlayerObject.js:
```diff
@@ -1,23 +1,133 @@
+        setRotationCount(previousCount => previousCount + 1);
+        setTimeout(() => setMotionIndex(1), 180);
+
+        for (let i = 0; i < 18; i++) {
```
위의 반복은 i가 20 미만인데 왜 여기부턴 18인가요?

j25nkh replied:
플레이어 오브젝트의 포지션을 변경하는 로직은 25ms 간격으로 20번 움직여, 좌표 1칸을 0.5초동안 가는 로직입니다.
반면 로테이션 (머리 방향) 값을 결정하는 로직은 한번에 5도씩 18번 움직여 90도 회전하는 로직이라 반복수가 다르게 설정되었습니다.

j25nkh committed:
fix: 플레이어 오브젝트 키컨트롤 로직 분리

- 키보드 입력값에 따라 플레이어 오브젝트의 포지션, 방향 및 애니메이션 모션을 결정하는 로직을 별도의 파일로 분리

sageherb reviewed (approved):
고생하셨습니다 👏

yourchaee reviewed (approved):
수고하셨습니다!!
